src/simulation/m_boundary_conditions.fpp

```diff
@@ -0,0 +1,1440 @@
+                                        mv(l, n + j, k, q, i) = &
+                                            mv(l, n , k, q, i)
+                                    end do
+                                end do
+                            end do
+                        end do
+                    end do
+                end if   
+
+            end if
+
+        !< z-direction =========================================================
+        elseif (bc_dir == 3) then !< z-direction
+
+            if (bc_loc == -1) then !< bc_z%beg
+
+                !$acc parallel loop collapse(4) gang vector default(present)
+                do i = 1, sys_size
+                    do j = 1, buff_size
+                        do l = -buff_size, n + buff_size
+                            do k = -buff_size, m + buff_size
+                                q_cons_vf(i)%sf(k, l, -j) = &
+                                    q_cons_vf(i)%sf(k, l, 0)
+                            end do
+                        end do
+                    end do
+                end do
+
+                if(qbmm .and. .not. polytropic) then
+                    !$acc parallel loop collapse(5) gang vector default(present)
+                    do i = 1, nb
+                        do j = 1, buff_size
+                            do l = -buff_size, n + buff_size
+                                do k = -buff_size, m + buff_size
+                                    do q = 1, nnode
+                                        pb(k, l, -j, q, i) = &
+                                            pb(k, l, 0, q, i)
+                                        mv(k, l, -j, q, i) = &
+                                            mv(k, l, 0, q, i)
+                                    end do
+                                end do
+                            end do
+                        end do
+                    end do
+                end if
+
+            else !< bc_z%end
+
+                !$acc parallel loop collapse(4) gang vector default(present)
+                do i = 1, sys_size
+                    do j = 1, buff_size
+                        do l = -buff_size, n + buff_size
+                            do k = -buff_size, m + buff_size
+                                q_cons_vf(i)%sf(k, l, p + j) = &
+                                    q_cons_vf(i)%sf(k, l, p)
+                            end do
+                        end do
+                    end do
+                end do
+
+                if(qbmm .and. .not. polytropic) then
+                    !$acc parallel loop collapse(5) gang vector default(present)
+                    do i = 1, nb
+                        do j = 1, buff_size
+                            do l = -buff_size, n + buff_size
+                                do k = -buff_size, m + buff_size
+                                    do q = 1, nnode
+                                        pb(k, l, p+j, q, i) = &
```

Comment:
some of these loop orderings don't make sense to me, should they not go from outer to inner index, so in this case the loops would go from outer to inner as k, l, p, q, i?

Reply:
@anandrdbz I believe what I have here is the same as what was in your initial PR for this feature.

Reply:
perhaps, but still requires explaining (some of the other loops also have this issue)

Reply:
I think these kernels weren't optimized to the fullest because they are orders of magnitude faster than the critical points as they only loop through two of the three dimensions at most. But yes, this might show a slight benefit by either using the q loop outside or adding a '!$acc loop seq' in front of the q while keeping the same order.

Reply:
However, it should make no meaningful difference to the overall run time for a single time step

Reply:
These have been reordered outer to inner with a 0.13% performance improvement for the test case I ran.

Reply:
understood, what @anandrdbz says makes sense.